precompiles/gmp/src/lib.rs

```diff
@@ -187,22 +189,75 @@ where
 			.map_err(|_| revert("Amount overflows balance"))?;
 
 		log::debug!(target: "gmp-precompile", "sending XCM via xtokens::transfer...");
-		let call: orml_xtokens::Call<Runtime> = match user_action {
-			VersionedUserAction::V1(action) => orml_xtokens::Call::<Runtime>::transfer {
-				currency_id,
-				amount,
-				dest: Box::new(action.destination),
-				dest_weight_limit: WeightLimit::Unlimited,
-			},
+		let call: Option<orml_xtokens::Call<Runtime>> = match user_action {
+			VersionedUserAction::V1(action) => {
+				log::debug!(target: "gmp-precompile", "Payload: V1");
+				Some(orml_xtokens::Call::<Runtime>::transfer {
+					currency_id,
+					amount,
+					dest: Box::new(action.destination),
+					dest_weight_limit: WeightLimit::Unlimited,
+				})
+			}
+			VersionedUserAction::V2(action) => {
+				log::debug!(target: "gmp-precompile", "Payload: V2");
+				// if the specified fee is more than the amount being transferred, we'll be nice to
+				// the sender and pay them the entire amount.
+				let fee = action.fee.min(amount_transferred);
```

Comment:
What actually prevents the relayer to set a super high fee? is it because the relayers work as consensus so they kind of all agree with the fees ?

Reply:
Relayers don't explicitly set any fees, and anyone can be a relayer. If no one likes the fees that a user provides in this payload, then it is likely no one will relay for them -- but the user can always do it themselves.

We will need to do some sort of documentation on how to best set a fee...

Reply:
I added a `Design` blurb in the main description to elaborate on this point, by the way.